napari/layers/_vectors_layer/model.py

```diff
@@ -0,0 +1,457 @@
+
+from typing import Union
+
+import numpy as np
+import scipy.signal as signal
+
+from .._base_layer import Layer
+from .._register import add_to_viewer
+from ..._vispy.scene.visuals import Line as LinesNode
+from ...util.event import Event
+from vispy.color import get_color_names
+
+from .view import QtVectorsLayer
+
+@add_to_viewer
+class Vectors(Layer):
+    """
+    Vectors layer renders lines onto the image.
+    There are currently two data modes supported:
+        1) image-like vectors where every pixel defines its own vector
+        2) coordinate-like vectors where x-y position is not fixed to a grid
+    Supports ONLY 2d vector field
+
+    Properties
+    ----------
+    vectors : np.ndarray
+        array of shape (N,4) or (N, M , 2)
+
+    data : np.ndarray
+        ABC implementation.
+        returns vectors property setter/getter
+
+    _original_data : np.ndarray
+        Used by averaging and length adjustments.
+        Is set during layer construction.
+        Is NOT updated if data is adjusted by assigning the 'vectors' property
+
+    averaging : int
+        (int, int) kernel over which to convolve the data
+        averaging.setter adjusts the underlying data
+        subscribe an observer by registering it with "averaging_bind_to"
+
+    width : int
+        width of the line in pixels
+
+    length : int or float
+        length of the line
+        length.setter adjusts the underlying data
+        subscribe an observer by registering it with "length_bind_to"
+
+    color : str
+        one of "get_color_names" from vispy.color
+
+    mode : str
+        control panel mode
+
+    Attributes
+    ----------
+    Private attributes
+    -------
+    _data_types
+    _data_type
+    _width
+    _color
+    _colors
+    _averaging
+    _length
+    _need_display_update
+    _need_visual_update
+    _raw_dat
+    _original_data
+    _current_data
+    _vectors
+
+    Public attributes
+    -------
+    name
+
+    See vispy's line visual docs for more details:
+    http://api.vispy.org/en/latest/scene.html
+    vispy.scene.visuals.Line
+    """
+
+    def __init__(self,
+                 vectors,
+                 width=1,
+                 color='red',
+                 averaging=1,
+                 length=1,
+                 name=None):
+
+        visual = LinesNode(antialias=True)
+        super().__init__(visual)
+
+        # events for non-napari calculations
+        self.events.add(emit_len=Event,
```

Comment:
not a big fan of the `emit_` names. would just use what they correspond to, i.e. `length` and `average`

Reply:
i brought this up in #182  -- that self.events is actually an EmitterGroup.

But seeing this: https://github.com/vispy/vispy/blob/master/vispy/util/event.py

It looks like vispy likes to define as such:
EmitterGroup  EventEmitter
                 |       |
Canvas.events.mouse_press
Canvas.events.resized
Canvas.events.key_press

which is weird to me because the Event itself is one level lower than the EventEmitter... but ok, I can switch back.